### ISSUE TITLE
[$500] mweb - Change Password  - If user get incorrect current password message, the field does not highlight

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#10523]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Access staging.new.expensify.com
2. Sign into a valid account
3. Tap on Profile icon > Security > Change Password
4. Enter incorrect password in the "Current password" field
5. Enter a viable password in the "New Password" field

## Expected Result:
The user expects to receive a message "Your password changes could not be saved because the old password entered is incorrect" AND that the current password field highlights with a red outline

## Actual Result:
The user only receives the error message, the field does not highlight as all

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** v1.1.89-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** any
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/186430532-c90c2d00-0003-43ce-a648-109b1b0feb53.mp4

<img src="https://user-images.githubusercontent.com/43996225/186430542-8b0a4e35-9139-406b-85c3-004e3570dbc1.png" width="40%" height="40%">

**Upwork job URL:** https://www.upwork.com/jobs/~0112662b9889b3b988

**Issue reported by:** Applause internal team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by michaelhaxhiu** [comment]
Job is on upwork - https://www.upwork.com/jobs/~01058690471e8c9ce2

**Comment by mmjindal2555** [comment]
RCA:
 - The password fields are not getting highlighted based on the form submit ,i.e., on error response of Auth API that validates current password.
Proposal: 
 - on change of Props(after receiving API response), if there is an error referring to the "old password" -> we can update a state variable which will help us to highlight the "Current Password" text field.
 
 Proposal - 
```
  this.state = {
            currentPassword: '',
            newPassword: '',
            errors: {
                currentPassword: false,
                newPassword: false,
                newPasswordSameAsOld: false,
            },
            incorrectCurrentPasswordError: false,
        };
```
 
```
    static getDerivedStateFromProps(nextProps) {
        return nextProps.account.error.includes('old password') ? {incorrectCurrentPasswordError: true} : {};
    }
```
```
                                        hasError={this.state.errors.currentPassword || this.state.incorrectCurrentPasswordError}
                                        errorText={this.state.errors.currentPassword && this.props.account.error !== ''
                                            ? this.props.account.error
                                            : this.getErrorText('currentPassword')}
```
![Screenshot 2022-09-06 at 9 52 42 PM](https://user-images.githubusercontent.com/14273484/188686892-ed422ec5-4b56-492d-955f-c0385e9086e6.png)

**Comment by michaelhaxhiu** [comment]
Thanks @parasharrajat. Is this a regression btw? Let's cc the design team after we address that!

**Comment by michaelhaxhiu** [comment]
cc @Expensify/design for input on this issue, which rajat mentioned above:
>I think we have tweaked this a page couple of times. A better approach would be to get the design team in and decide on the look and feel of this page before we make any changes.

**Comment by shawnborton** [comment]
Yeah, I think Rajat is referencing this issue here: https://github.com/Expensify/App/issues/10493#issuecomment-1258544863

But regardless, I think we highlight the field that has an error with a red outline and an error message right below it.

**Comment by rs181915** [comment]
> RCA:
> 
> * The password fields are not getting highlighted based on the form submit ,i.e., on error response of Auth API that validates current password.
>   Proposal:
> * on change of Props(after receiving API response), if there is an error referring to the "old password" -> we can update a state variable which will help us to highlight the "Current Password" text field.
> 
> Proposal -
> 
> ```
>   this.state = {
>             currentPassword: '',
>             newPassword: '',
>             errors: {
>                 currentPassword: false,
>                 newPassword: false,
>                 newPasswordSameAsOld: false,
>             },
>             incorrectCurrentPasswordError: false,
>         };
> ```
> 
> ```
>     static getDerivedStateFromProps(nextProps) {
>         return nextProps.account.error.includes('old password') ? {incorrectCurrentPasswordError: true} : {};
>     }
> ```
> 
> ```
>                                         hasError={this.state.errors.currentPassword || this.state.incorrectCurrentPasswordError}
> ```
> 
> ![Screenshot 2022-09-06 at 9 52 42 PM](https://user-images.githubusercontent.com/14273484/188686892-ed422ec5-4b56-492d-955f-c0385e9086e6.png)

we can also use states this way...

  this.state = {
      currentPassword: '',
      newPassword: '',
      errors: {
        currentPassword: {
          status: true /* false */,
          message: 'set error from validations for currentPassword'
        },
        newPassword: {
          status: true /* false */,
          message: 'set error from validations for newPassword'
        },
      },
    };
    or this way...
       this.state = {
      currentPassword: {
        value: '',
        isError: false,
        errorMessage: '',
      },
      newPassword: {
        value: '',
        isError: false,
        errorMessage: '',
      },
    };
    
    i'm primarily involved in debugging & logical stuff... 
    and I can resolve this issue... issue doesn't seem compacted for me. 
   
   
   It's my first proposal on Expensify...
   so yea, I might need 1-2 days to set up the dev environment...
    but after that, I can submit resolved issue pr just in a day.

**Comment by michaelhaxhiu** [comment]
fyi - This isn't a regression so no need to include `RCA` in your proposal. @parasharrajat can you review this today?

**Comment by parasharrajat** [comment]
OK, So what is finalized here? 

1. make the input outline red.
2. Show the error below the input.

2 can't be done as the error is coming from the backend. As the backend is returning the error as a generic error, there is no go to way to identify that the error is related to old password field. 

But as shown in the proposal https://github.com/Expensify/App/issues/10523#issuecomment-1238374969, we can search for `old password` text in the error to identify it as a error related to old password field. 


Currently, I don't know the correct way to handle OpenAPI errors. @flodnv Could you please help me with this part? Happy to discuss on DM or slack channels. Thank you.

---

@rs181915 Could you please format your https://github.com/Expensify/App/issues/10523#issuecomment-1264187387 proposal so that it is readable? 

Links: 
- [CONTRIBUTING Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md)
- [App overview and local development](https://github.com/Expensify/App#local-development)
- [CODE OF CONDUCT](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md)
- [Code style and patterns](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md)
- [Styling Components](https://github.com/Expensify/App/blob/main/contributingGuides/STYLING.md)

I would recommend checking out some previous jobs, which already got a contributor assigned to see how they posted the proposals. There is also [this part](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md?rgh-link-date=2022-07-19T09%3A31%3A19Z#propose-a-solution-for-the-job) in our contributor guidelines which you might find useful.

Please, also share the exact changes in code you had to make to fix this issue. You can see from all the other issues and contributors that we won't just take the code and run away, no need to worry about that :)

Thank you!

**Comment by flodnv** [comment]
Posted here: https://expensify.slack.com/archives/C01GTK53T8Q/p1665142427568719

@parasharrajat there was also an earlier proposal from @mmjindal2555 https://github.com/Expensify/App/issues/10523#issuecomment-1238374969

**Comment by iwiznia** [comment]
Coming from https://expensify.slack.com/archives/C01GTK53T8Q/p1665142427568719
TLDR:
- We already have a way of doing this. The server needs to send onyx data setting `errorFields` in the object like [this](https://github.com/Expensify/Web-Expensify/blob/9d8c5fe80bfa058f5dfb54f494b39ac7502be59e/lib/Error/WalletAdditionalDetails/VerificationError.php#L132)
- The frontend part is already handled [here](https://github.com/Expensify/App/blob/60b381c5432905c1de27812bf01fb1e420acece2/src/libs/Middleware/SaveResponseInOnyx.js#L27-L29)
- This needs to be worked on internally (at least the server part, but probably better if it is both parts)

**Comment by mmjindal2555** [comment]
@flodnv how are you guys planning to take it further? I see the issue is not 'External' now, does it mean it won't be assigned to any external contributor? 
i proposed the solution which was summarised by @parasharrajat in [this comment](https://github.com/Expensify/App/issues/10523#issuecomment-1270528529), it involve changes in frontend side only. 
If required I can also explore and share a solution for server side changes. This is my first proposal and I plan to work with expensify continuously.

**Comment by flodnv** [comment]
Hi @mmjindal2555. Indeed, the ideal for this is API side, which is unfortunately not (yet) accessible to open source contributors. We value your contributions and are looking forward to your next proposals.

**Comment by flodnv** [comment]
No update, this week I've been swamped by other daily priorities.
